lib/elixir/lib/enum.ex

```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:error, _module} ->
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
```

Comment:
This change is backwards incompatible. It is ok for a module to return ANOTHER module which has the proper implementation. This change also defeats the whole point of returning `{:error, __MODULE__}`, which is to avoid two protocol dispatches.